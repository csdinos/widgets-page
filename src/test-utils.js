import React from 'react'
import {render as rtlRender} from '@testing-library/react'
import {createStore} from 'redux'
import {Provider} from 'react-redux'
import {reducer} from './reducers'

function render(
  ui,
  {
    initialState,
    store = createStore(reducer, initialState),
    ...renderOptions
  } = {}
) {
  function Wrapper({children}) {
    return <Provider store={store}>{children}</Provider>
  }

  return rtlRender(ui, {wrapper: Wrapper, ...renderOptions})
}

export const initialState = {
  widgets: [],
  newWidget: {
    id: -1,
    name: '',
    language: '',
  },
  deleteModalId: null,
  hasLoaded: false,
  shouldUpdateStorage: false
}

export * from '@testing-library/react'
export {render}