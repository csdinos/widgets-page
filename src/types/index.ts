export type Widget = {
  id: number;
  name: string;
  language: string;
}
export type State = {
  widgets: Array<Widget>
  newWidget: Widget
  deleteModalId: number | null,
  hasLoaded: boolean,
  shouldUpdateStorage: boolean
}