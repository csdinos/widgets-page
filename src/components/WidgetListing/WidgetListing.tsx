import {Button, Grid, Typography, Card, CardContent, CardActions} from '@material-ui/core'
import {useDispatch, useSelector} from 'react-redux'
import {State} from '../../types'
import * as actions from '../../actions'
import {useEffect} from 'react'

export const WidgetListing = () => {
  const widgets = useSelector((state: State) => state.widgets)
  const hasLoaded = useSelector((state: State) => state.hasLoaded)
  const shouldUpdateStorage = useSelector((state: State) => state.shouldUpdateStorage)
  const dispatch = useDispatch()

  const loadWidgetsFromLocalStorage = () => {
    const widgets = localStorage.getItem('widgets');

    if (widgets) {
      dispatch(actions.add(JSON.parse(widgets)))
    }
  }

  const updateLocalStorage = () => {
    localStorage.setItem('widgets', JSON.stringify(widgets))
  }

  useEffect(() => {
    if (!hasLoaded) {
      loadWidgetsFromLocalStorage()

      dispatch(actions.setHasLoaded(true))
    }
  },[])

  useEffect(() => {
    if (shouldUpdateStorage) {
      updateLocalStorage()
      dispatch(actions.setShouldUpdateStorage(false))
    }
  },[shouldUpdateStorage])

  return (
    <>
      {
        widgets.map((w, i) => (
          <Grid item xs={6} key={i}>
            <Card variant="outlined" className="widget-card">
              <CardContent>
                <Typography variant="body1" gutterBottom>Name: {w.name}</Typography>
                <Typography variant="body1" gutterBottom>Language: {w.language}</Typography>
              </CardContent>
              <CardActions>
                <Button
                  className="delete-btn"
                  fullWidth
                  variant="contained"
                  color="secondary"
                  onClick={() => {
                    dispatch(actions.setIdForDeleteModal(w.id))
                  }}
                >Delete</Button>
              </CardActions>
            </Card>
          </Grid>
        ))
      }
    </>
  )
}