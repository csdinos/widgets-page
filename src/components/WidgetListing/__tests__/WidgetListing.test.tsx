import React from "react";
import { render, initialState } from "../../../test-utils";
import {WidgetListing} from "../WidgetListing";
import {State} from '../../../types'

const state: State = {
  ...initialState,
  widgets: [{
    name: 'Doge',
    language: 'Meme',
    id: 1994
  }],
}

describe("<WidgetListing />", () => {
  it("When we have initial empty state, it should display a an empty widget list", async () => {
    const { asFragment } = render(<WidgetListing />);
    expect(asFragment()).toMatchSnapshot()
  });

  it("When we have a filled state with one widget, it should display a widget list with one element", async () => {
    const { asFragment } = render(<WidgetListing />, {initialState: state});
    expect(asFragment()).toMatchSnapshot()
  });
});