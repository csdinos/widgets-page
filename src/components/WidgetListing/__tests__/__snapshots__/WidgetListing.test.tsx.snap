// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WidgetListing /> When we have a filled state with one widget, it should display a widget list with one element 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
  >
    <div
      class="MuiPaper-root MuiCard-root widget-card MuiPaper-outlined MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
        >
          Name: Doge
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
        >
          Language: Meme
        </p>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained delete-btn MuiButton-containedSecondary MuiButton-fullWidth"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Delete
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<WidgetListing /> When we have initial empty state, it should display a an empty widget list 1`] = `<DocumentFragment />`;
